### ISSUE TITLE
Update slot value with entity value from Rasa NLU only after verifying if it is in the list of candidates

### ISSUE THREAD

**Issue by msamogh** [RasaHQ/rasa_core#257]
This is relevant (as specified in #254) when a slot is defined with a set of possible values in the domain file, like below.

```yaml
slots:
  expenses_category:
    type: categorical
    values:
      - utility
      - travel
      - investment
      - grocery
      - fuel
      - education
      - dining
```

Rather than blindly update the slot value with that of the entity, a check should be performed if the entity value belongs to the set of candidate values.

Perhaps, raise a warning or exception?

**Comment by nachoaguadoc** [comment]
I believe a warning or exception is not an adequate solution to the problem if the value is still getting assigned. If categorical slots can get updated independently of the value, what's the point of using them instead of just a text slot? If the performed check indicates that the proposed value is not between the set of candidates, maybe a warning can be issued as debug information, but in no case the slot content should be updated

**Comment by msamogh** [comment]
One use I could think of is a pattern that benefits from having pre-defined values for the slot most of the time, but if it's not one of them, you might want an "none of these" case. 
As to how it's different from just having a separate "None of these" category, you have a loss of information there because you don't know the value.

**Comment by nachoaguadoc** [comment]
For the case you propose I believe it's cleaner to have a text slot and check the values manually in an Action. There you can choose, depending on the value, whether you want to update an unfeaturized slot of your own with the given value or "none-of-the-above". I still think that the main point of having the categorized slot is that its value has to be one of the categories.

In any case, let's see what the devs think that should be done about it

**Comment by tmbo** [comment]
The difference between categorical and text slot is, that the value set in a categorical slot will influence the prediction. E.g. if you specify 3 different values for your categorical slot (`eggs`, `milk`, `ham`) there are going to be *four different featurizations*. One for each value and one for no value set. If you set a value that is not part of the listed values, to the ML model it will look like no value is set.

**Comment by tmbo** [comment]
in #263 I added a logger output to warn if this happens.
